gnucash/import-export/aqb/assistant-ab-initial.c

```diff
@@ -332,11 +396,14 @@ aai_on_finish (GtkAssistant *assistant, gpointer user_data)
     ABInitialInfo *info = user_data;
     Account *root;
 
-    g_return_if_fail(info && info->gnc_hash);
+    g_return_if_fail(info && info->gnc_hash && info->gnc_revhash);
 
+    /* Remove GnuCash accounts from reverse hash table which are still
+     * matched to an AqBanking account. For the remaining GnuCash accounts
+     * the KVPs must be cleared (i.e. deleted) */
+    g_hash_table_foreach(info->gnc_hash, (GHFunc) remove_acc_cb, info->gnc_revhash);
```

Comment:
OK. looks better :-)

Reply:
After rethinking your suggestion I come to the conclusion that it's not equivalent to my implementation. Let's make the following example:

On entrance to the account matching page `info->hash` matching AqBanking accounts (ab_acc) with GnuCash accounts (gnc_acc) may look like

ab_acc | gnc_acc
--- | ---
a1 | -
a2 | g2
a3 | g1
a4 | g3

This would result in `info->revhash` look like

gnc_acc | ab_acc
--- | ---
g1  | a3
g2  | a2
g3  | a4

Let's assume that the user now will match g1 with a1, g4 with a4 (please note that a3 is unmatched after the first operation). This would finally result in the following `info->hash` on exit of the account matching page:

ab_acc | gnc_acc
--- | ---
a1 | g1
a2 | g2
a3 | -
a4 | g4

As a result of this g1 and g2 shall be removed from `info->revhash`:

gnc_acc | ab_acc
--- | ---
~~g1~~  | ~~a3~~
~~g2~~  | ~~a2~~
g3  | a4

That means that there is only g3 remaining in `info->revhash`, which is now no longer matched with an AqBanking account and whose KVPs need to be deleted.

As far as I can see your suggestion will iterate through `info->revhash` and will lookup matches in `info->hash` by the AqBanking accounts a3, a2, and a4. The callback function `in_forward_hash` will return `TRUE` only for a2 and a4 and thus will cause a removal of g2 and g3 from `info->revhash`, and this will cause deletion of the KVPs for g1 but will keep the KVPs for g3.

I think I will keep the code as it is and may add further comments. Otherwise `in_forward_hash` should lookup `info->hash` by value, which seems to be more inefficient to me.

Reply:
OIC. That can be corrected by changing the condition in `in_forward_hash` to
`if (gnc_hash_lookup_value((GHashTable*)user_data, value) == key)` *but* this approach depends on *not* using the `g_hash_table_insert` optimization of passing `key` to both `key` and `value` because this way you do need `ab_acc` in `revhash`.

Reply:
Strike that, checking that `g_hash_lookup((GHashTable*)user_data, value) == key` would also exclude g1 as that's changed. `remove_acc_cb` modified as I suggest is indeed the right way.